src/extensions/default/InlineBorderRadiusEditor/BorderRadiusEditor.js

```diff
@@ -0,0 +1,716 @@
+        var unit = event.target.parentNode.id.split("-");
+        this.updateRadios("tr",unit[2]);
+    };
+
+    BorderRadiusEditor.prototype._handleBRRadioChange = function(event){
+        this._clearAllRadio("br");
+        this.$element.find("#"+event.target.parentNode.id).addClass("selected");
+        var unit = event.target.parentNode.id.split("-");
+        this.updateRadios("br",unit[2]);
+    };
+
+    BorderRadiusEditor.prototype._handleBLRadioChange = function(event){
+        this._clearAllRadio("bl");
+        this.$element.find("#"+event.target.parentNode.id).addClass("selected");
+        var unit = event.target.parentNode.id.split("-");
+        this.updateRadios("bl",unit[2]);
+    };
+
+    BorderRadiusEditor.prototype._handleALLRadioChange = function(event){
+        this._clearAllRadio("all");
+        this.$element.find("#"+event.target.parentNode.id).addClass("selected");
+        var unit = event.target.parentNode.id.split("-");
+        this.updateRadios("all",unit[2]);
+    };
+
+
+    BorderRadiusEditor.prototype._handleTRCHange = function() {
+        var self = this;
+        var newValue = this.$trslider.val().trim();
+        _handleChanges.call(self, this.$trslider, "TR", newValue);
+    };
+
+    BorderRadiusEditor.prototype._handleBLCHange = function() {
+        var self = this;
+        var newValue = this.$blslider.val().trim();
+        _handleChanges.call(self, this.$blslider, "BL", newValue);
+    };
+
+    BorderRadiusEditor.prototype._handleBRCHange = function() {
+        var self = this;
+        var newValue = this.$brslider.val().trim();
+        _handleChanges.call(self, this.$brslider, "BR", newValue);
+    };
+
+    BorderRadiusEditor.prototype._handleALLCHange = function() {
+        var self = this;
+        var newValue = this.$allCornerSlider.val().trim();
+        _handleChanges.call(self, this.$allCornerSlider, "ALL", newValue);
+    };
+
+    /**
+    * Global handler for keys in the borderradius editor. Catches undo/redo keys and traps
+    * arrow keys that would be handled by the scroller.
+    */
+    BorderRadiusEditor.prototype._handleKeydown = function (event) {
+        var hasCtrl = (brackets.platform === "win") ? (event.ctrlKey) : (event.metaKey);
+        if (hasCtrl) {
+            switch (event.keyCode) {
+                case KeyEvent.DOM_VK_Z:
+                    if (event.shiftKey) {
+                        this.redo();
+                    } else {
+                        this.undo();
+                    }
+                    return false;
+                case KeyEvent.DOM_VK_Y:
+                    this.redo();
+                    return false;
+            }
+        }
+    };
+
+    BorderRadiusEditor.prototype._commitChanges = function(value) {
+        var result="";
+        var _array = value.split(" ");
+        for(var i=0;i<_array.length;i++){
+            result+=_array[i];
+        }
+        this._values = result;
+        this._callback(value);
```

Comment:
Does this mean you'll call the callback more than once, since `_commitChanges` is called more than once above?  If so, is that expected?  It's an odd pattern: usually you'd just call a callback once, and if you need to call it many times, you'd do some kind of event handler instead.

Reply:
@humphd In this case, the event handler is being implemented in the InlineBorderRadiusEditor.js layer, whenever the UI element(slider) changes, it will trigger the event handler in BorderRadiusEditor.js, then BorderRadiusEditor.js will update the UI and then send control back to InlineBoderRadiusEditor.js by calling the _callback(), which is mapping to the _handleBorderRadiusChange() in the InlineBorderRadius.js.  the _handleBorderRadiusChange() will be updating document text.

Reply:
@humphd basically, in my code, the InlineBorderRadiusEditor.js handles the Document updating tasks while BorderRadiusEditor.js handles UI Updating first then send control to InlineBorderRadiusEditor.js to update the Text in the hostEditor.

Reply:
@humphd I'm not really sure how to implement the event handler for the _callback, since BorderRadiusEditor and InlineBorderRadiusEditor works on different layers, could you give more hint on this? or maybe we can discuss about this in class tomorrow.